### PR TITLE
Message bus code merge

### PR DESCRIPTION
**Pull request by suryakumar1024**
Moved messagebus code to py-utils/src/utils/msg_bus

This branch code has Kafka-python adaptee and Confluent-kafka-adaptee configured codes.

### CODE REVIEW

File: py-utils/src/utils/msg_bus/analytical_example.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+from message_bus.bus import Bus
+# from bus.callback import MyCallback
+from message_bus import Producer
+from message_bus import Consumer
+from message_bus.analytical_consumer import AnalyticalConsumer
+# from config import Config
+# from bus.topic import Topic
+from message_bus.message import Message
+
+from message_bus.utils import log_decorator
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/analytical_example.py]: Please reuse already existing logger in the cortx-utils

File: py-utils/src/utils/msg_bus/analytical_example.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+from message_bus.bus import Bus
+# from bus.callback import MyCallback
+from message_bus import Producer
+from message_bus import Consumer
+from message_bus.analytical_consumer import AnalyticalConsumer
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/analytical_example.py]: Please separate analytical consumer into different patch

File: py-utils/src/utils/msg_bus/analytical_example.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from message_bus.message import Message
+
+from message_bus.utils import log_decorator
+
+
+
+@log_decorator
+def main():
+    # config = Config()
+    bus_start = Bus('kafka')
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/analytical_example.py]: Bus -> MessageBus

File: py-utils/src/utils/msg_bus/analytical_example.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+from message_bus.bus import Bus
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/analytical_example.py]: Move all the test / example code to test folder. py-utils/test/message_bus/

File: py-utils/src/utils/msg_bus/.gitignore [hunk 5]
```diff
@@ -0,0 +1,4 @@
+*.pyc
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/.gitignore]: Please rename old message_bus to "msg_bus" and move this code to "message_bus" folder.

File: py-utils/src/utils/msg_bus/analytical_example.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    print('Receiving messages 2 . . .')
+    for each in msg2:
+        print(each)
+
+
+if __name__ == "__main__":
+    main()
+
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/analytical_example.py]: Please remove extra empty lines.

File: py-utils/src/utils/msg_bus/benchmarks/kafka_python_vs_confluent.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
+from __future__ import absolute_import
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/benchmarks/kafka_python_vs_confluent.py]: Move this code to cortx-experiments repo

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/msg_bus/confluent_kafka_example.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
+from __future__ import absolute_import
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/confluent_kafka_example.py]: Please move it to py-utils/test/message_bus/

File: py-utils/src/utils/msg_bus/kafka_python_example.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
+from __future__ import absolute_import
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/kafka_python_example.py]: Please move it to py-utils/test/message_bus/

File: py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py]: Lets move this file to message_bus itself. Also Please rename to file and class to "MessageBroker" which is the base class for All the Message Brokers such as Kafka to be written

File: py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+from confluent_kafka.cimpl import Producer, Consumer
+from confluent_kafka.admin import AdminClient, NewTopic, ClusterMetadata
+from message_bus.adaptee import Adaptee
+from message_bus.utils import log_decorator
+
+ConsumerRecord = namedtuple("ConsumerRecord",
+                            ["topic", "message", "partition", "offset", "key"])
+
+
+class ConfluentAdaptee(Adaptee):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py]: Rename this to "KafkaMessageBroker". Fundamentally its MessageBroker.

File: py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+class Adaptee(object):
+    def __init__(self):
+        pass
+
+    def send(self):
+        pass
+
+    def receive(self):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py]: Please add parameters to the interfaces, even though its base abstract class

File: py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+    def __init__(self, config):
+        try:
+            self.config = config
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py]: Please names like _config or __config for all the protected or private members as conventions.

File: py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+            self.mapper = {}
+        except Exception as e:
+            print(e)
+
+    def create_admin(self):
+        config = self.config['client'][0]
+        # self.admin = AdminClient(**config)
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py]: Remove all the commented code once done. Until this is done, patch can not be merged.

File: py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+class ConfluentAdaptee(Adaptee):
+
+    def __init__(self, config):
+        try:
+            self.config = config
+            self.mapper = {}
+        except Exception as e:
+            print(e)
+
+    def create_admin(self):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py]: Is this a public interface ? If not Please prefix as _xxx() or __xxx()

File: py-utils/src/utils/msg_bus/message_bus/adaptee/kafka_python_adaptor.py [hunk 16]
```diff
@@ -0,0 +1,69 @@
+from kafka import KafkaAdminClient, KafkaProducer, KafkaConsumer
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/adaptee/kafka_python_adaptor.py]: Lets use either confluent or kafka python. We have reviewed the comparison and we will only use one of them. WE need to limit about adding 3rd party dependencies in CORTX code.

File: py-utils/src/utils/msg_bus/message_bus/admin.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
+
+from bus.client import BusClient
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/admin.py]: What is the purpose of this class ? Please remove or move to separate patch if not part of phase 1.

File: py-utils/src/utils/msg_bus/message_bus/analytical_consumer.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
+from message_bus import Consumer
+from message_bus.utils import add, check_topic, last_one_hour_data
+
+class AnalyticalConsumer(Consumer):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/analytical_consumer.py]: Please move this into separate patch. Not a phase 1 deliverable.

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+from message_bus.adaptee import KafkaAdaptee
+from message_bus.bus.callback import MyCallback
+from message_bus.template import kafkaFactory, ConfluentFactory
+from message_bus.template import Factory, Singleton
+from message_bus.bus.topic_schema import TopicSchema
+from message_bus.bus.topic import Topic
+from message_bus.utils import log_decorator
+
+
+class Bus(metaclass=Singleton):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: Please rename to MessageBus.

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py [hunk 20]
```diff
@@ -0,0 +1,142 @@
+import logging
+log = logging.getLogger(__name__)
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: Remove all the logging code and Use already existing logger from cortx-utils

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+            "confluent-kafka": ConfluentFactory
+        })
+        self.__load_adaptor(mq_bus_name)
+        if self.config is not None:
+            self.__load_topic()
+
+        self.schema = TopicSchema()
+        self.client_list = []
+
+    def __load_adaptor(self, mq_bus_name):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: Please rename all the adaptor to adapter. The meaning are different. In our context we need adapter, I think.

File: py-utils/src/utils/msg_bus/kafka_python_example.py [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+from message_bus.bus.bus_frame import Bus
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/kafka_python_example.py]: Please add copyright notice to all the files. Refer to the other cortx-utils files for the format. Copy/Paste from there.

File: py-utils/src/utils/msg_bus/message_bus/bus/callback.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+        super().post_subscribe(consumer, topic, pattern=None, listener=None)
+
+    def pre_receive(self,consumer):
+        super().pre_receive(consumer)
+
+    def post_receive(self, consumer):
+        super().pre_receive(consumer)
+
+
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/bus/callback.py]: Remove any extra lines.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/msg_bus/message_bus/config.py [hunk 24]
```diff
@@ -0,0 +1,33 @@
+
+class Config(object):
+    def __init__(self):
+        pass
+
+
+class KafkaConfig(Config):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/config.py]: For now, Conf class  from cortx-utils can be used. These values need to be supplied from config file.

File: py-utils/src/utils/msg_bus/singleton_example.py [hunk 25]
```diff
@@ -0,0 +1,52 @@
+from bus.bus_frame import Bus
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/singleton_example.py]: move to test/message_bus/

File: py-utils/src/utils/msg_bus/notifier_example.py [hunk 26]
```diff
@@ -0,0 +1,53 @@
+from message_bus.bus import Bus
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/notifier_example.py]: Move to py-utils/test/message_bus/

File: py-utils/src/utils/msg_bus/message_bus/utils.py [hunk 27]
```diff
@@ -0,0 +1,44 @@
+import logging
+import time
+from datetime import datetime, timedelta
+
+logging.basicConfig(filename='log_file.log', level=logging.DEBUG)
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/utils.py]: Please use cortx-utils logger. By default no logs should be generated only the debug logs.

File: py-utils/src/utils/msg_bus/message_bus/utils.py [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+        logging.debug(f"{func.__name__} function ended at{end} total time {end-start}")
+    return wrapper
+
+def validate_string(func):
+    def wrapper(*args, **kwargs):
+        for k, v in kwargs.items():
+            print(f"k {k}, v{v}")
+    return wrapper
+
+def add(messages):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/utils.py]: Please avoid and remove all the global functions. In fact this file itself needs to be removed. Move these to resp classes.

File: py-utils/src/utils/msg_bus/message_bus/template/singleton.py [hunk 29]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/template/singleton.py]: Move the template to outside of message_bus and separate patch. 
Best folder is py-utils/src/template/

File: py-utils/src/utils/msg_bus/message_bus/template/factory.py [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+
+class Factory(object):
+
+    def __init__(self, adaptors):
+        self.m_dict = adaptors
+
+    def __call__(self, mq_name):
+        return self.m_dict[mq_name]()
+
+class kafkaFactory(Factory):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/template/factory.py]: - This class should be called "MessageBrokerFactory". 
- Template should be generic and not have anything specific to any modules. Move this part of the code to message_bus/message_broker.py along with MessageBroker class.

File: py-utils/src/utils/msg_bus/message_bus/producer.py [hunk 31]
```diff
@@ -0,0 +1,24 @@
...
+        super().__init__(busHandle, 'PRODUCER')
+
+    def send(self, message):
+        log.info("Bus send message")
+        print("Bus send message")
+        super().send(message)
+        log.info("Bus send message complete")
+        print("Bus send message complete")
+
+    # Above implementation is enough
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/producer.py]: Please remove any code that is not required.

File: py-utils/src/utils/msg_bus/message_bus/message/message.py [hunk 32]
```diff
@@ -0,0 +1,29 @@
+from message_bus.constants import MessageConst as msg_c
+from message_bus.message import MessageHeader
+
+class Message(object):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/message/message.py]: Python3 does not require you to derive out of object. Please remove derivation.

File: py-utils/src/utils/msg_bus/message_bus/message/header.py [hunk 33]
```diff
@@ -0,0 +1,15 @@
+from message_bus.message import MessageConfig
+
+class MessageHeader(object):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/message/header.py]: Move this class to message.py itself.

File: py-utils/src/utils/msg_bus/message_bus/message/format.py [hunk 34]
```diff
@@ -0,0 +1,5 @@
+
+class MessageFormat(object):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/message/format.py]: Move this class to message.py itself.

File: py-utils/src/utils/msg_bus/message_bus/message/config.py [hunk 35]
```diff
@@ -0,0 +1,20 @@
+from message_bus.utils import validate_string
+from message_bus.constants import MessageConst as msg_c
+
+class MessageConfig(object):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/message/config.py]: Move this class to message.py itself.

File: py-utils/src/utils/msg_bus/message_bus/message/__init__.py [hunk 36]
```diff
@@ -0,0 +1,5 @@
+from __future__ import absolute_import
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/message/__init__.py]: All the code in this folder can be merged into message_bus/message.py

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py [hunk 37]
```diff
@@ -0,0 +1,142 @@
...
+        self.bus_callback.pre_receive(consumer)
+        consumer_obj = self.adaptor.receive(consumer)
+
+        if self.notifier is not None:
+            consumer_obj = self.notifier.get_caller(consumer_obj)
+
+        self.bus_callback.post_receive(consumer)
+        return consumer_obj
+
+    def subscribe(self, consumer, topic, notifier, pattern=None, listener=None):
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: Subscribe and unsubscribe was not part of the design. Please defer this to later patches. 

We need a working code without subscription. Subscription will / may be added later.

File: py-utils/src/utils/msg_bus/message_bus/bus/__init__.py [hunk 38]
```diff
@@ -0,0 +1,5 @@
+from __future__ import absolute_import
```
Comment (ujjwalpl) [inline py-utils/src/utils/msg_bus/message_bus/bus/__init__.py]: Lets consolidate this code from this folder into message_bus/message_bus.py

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/msg_bus/message_bus/message/format.py (hunk 34 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/message/format.py]: Merged

File: py-utils/src/utils/msg_bus/message_bus/message/header.py (hunk 33 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/message/header.py]: Merged

File: py-utils/src/utils/msg_bus/analytical_example.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/msg_bus/analytical_example.py]: Renamed Bus class to MessageBus

File: py-utils/src/utils/msg_bus/.gitignore (hunk 5 above)
Comment (suryakumar1024) [reply py-utils/src/utils/msg_bus/.gitignore]: Updated and now old code can be found in msg_bus new code in message_bus

File: py-utils/src/utils/msg_bus/singleton_example.py (hunk 25 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/singleton_example.py]: Moved

File: py-utils/src/utils/msg_bus/notifier_example.py (hunk 26 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/notifier_example.py]: Moved

File: py-utils/src/utils/msg_bus/message_bus/bus/callback.py (hunk 23 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/bus/callback.py]: Removed

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py (hunk 19 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: Renamed

File: py-utils/src/utils/msg_bus/kafka_python_example.py (hunk 9 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/kafka_python_example.py]: Moved

File: py-utils/src/utils/msg_bus/confluent_kafka_example.py (hunk 8 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/confluent_kafka_example.py]: Moved

File: py-utils/src/utils/msg_bus/analytical_example.py (hunk 6 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/analytical_example.py]: Removed

File: py-utils/src/utils/msg_bus/analytical_example.py (hunk 4 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/analytical_example.py]: Moved

File: py-utils/src/utils/msg_bus/message_bus/message/__init__.py (hunk 36 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/message/__init__.py]: merged

File: py-utils/src/utils/msg_bus/message_bus/message/message.py (hunk 32 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/message/message.py]: Removed

File: py-utils/src/utils/msg_bus/message_bus/producer.py (hunk 31 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/producer.py]: Removed

File: py-utils/src/utils/msg_bus/message_bus/bus/__init__.py (hunk 38 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/bus/__init__.py]: Consolidated

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py (hunk 21 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: Renamed

File: py-utils/src/utils/msg_bus/message_bus/template/singleton.py (hunk 29 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/template/singleton.py]: Moved

File: py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py (hunk 12 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py]: Params added

File: py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py (hunk 14 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/adaptee/confluent_kafka_adaptor.py]: Removed

File: py-utils/src/utils/msg_bus/message_bus/admin.py (hunk 17 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/admin.py]: BusClient is a abstract base class. Producer , Consumer are types of BusClient

File: py-utils/src/utils/msg_bus/message_bus/analytical_consumer.py (hunk 18 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/analytical_consumer.py]: Moved

File: py-utils/src/utils/msg_bus/message_bus/utils.py (hunk 28 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/utils.py]: Moved to notifier class

File: py-utils/src/utils/msg_bus/kafka_python_example.py (hunk 22 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/kafka_python_example.py]: copyright added

File: py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py (hunk 20 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/bus/bus_frame.py]: removed old loggers

File: py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py (hunk 10 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/message_bus/adaptee/adaptee.py]: moved

File: py-utils/src/utils/msg_bus/benchmarks/kafka_python_vs_confluent.py (hunk 7 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/benchmarks/kafka_python_vs_confluent.py]: moved

File: py-utils/src/utils/msg_bus/analytical_example.py (hunk 2 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/analytical_example.py]: moved

File: py-utils/src/utils/msg_bus/analytical_example.py (hunk 1 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/msg_bus/analytical_example.py]: removed log_decorator

**Comment by suryakumar1024** [comment]
Closing this PR. Since new PR raised directly in Seagate Repo
